[FLINK-19554][e2e] Implement a unified testing framework for connectors

PatrickRen opened the pull request:
## What is the purpose of the change

This pull request implements an end-to-end unified testing framework for connectors based on [Testcontainers](https://www.testcontainers.org/). An example test for Kafka connector is also provided in this PR. 

## Brief change log

- Implement connector testing framework 
- Implement an end-to-end test for Kafka connector

## Verifying this change

This change added tests and can be verified as follows:
- Make sure you have Docker environment on your machine
- Run integrated test using "mvn verify". This will start a Flink cluster with 1 JM and 1 TM, along with a Kafka broker on Docker, then run an end-to-end test for Kafka connector. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

becketqin commented on flink-end-to-end-tests/flink-connector-testing-framework/flink-connector-e2e-common/pom.xml:
```diff
@@ -0,0 +1,83 @@
+    <parent>
+        <artifactId>flink-connector-testing-framework</artifactId>
+        <groupId>org.apache.flink</groupId>
+        <version>0.1-SNAPSHOT</version>
```
This should follow the version of the project, which is 1.12-SNAPSHOT at this point.

becketqin commented on flink-end-to-end-tests/flink-connector-testing-framework/flink-connector-e2e-kafka/pom.xml:
```diff
@@ -0,0 +1,130 @@
+    <parent>
+        <artifactId>flink-connector-testing-framework</artifactId>
+        <groupId>org.apache.flink</groupId>
+        <version>0.1-SNAPSHOT</version>
```
The version here should be 1.12-SNAPSHOT.

becketqin commented on flink-end-to-end-tests/flink-connector-testing-framework/pom.xml:
```diff
@@ -0,0 +1,108 @@
+    <groupId>org.apache.flink</groupId>
+    <artifactId>flink-connector-testing-framework</artifactId>
+    <packaging>pom</packaging>
+    <version>0.1-SNAPSHOT</version>
```
The version should be 1.12-SNAPSHOT.

becketqin commented on flink-end-to-end-tests/flink-connector-testing-framework/flink-connector-e2e-kafka/src/test/resources/log4j2.properties:
```diff
@@ -0,0 +1,24 @@
+# limitations under the License.
+#
+
+rootLogger.level=INFO
```
Setting the default logging level to INFO may make the log too verbose. Can we change this to ERROR or only turn on the INFO logging for some specific loggers?

becketqin reviewed (commented):
@PatrickRen Thanks for the patch. I took a quick look and left some comments. 

Given that we would like to let developers use this testing framework for all the connectors, we probably need a top level README.md file to explain how to use the testing framework. At least we need a thorough Java doc.

Besides that, it looks that we can use the same framework for both IT cases and E2E tests. Some top level comments / suggestions about the abstraction to achieve that.

### The interfaces to bring up the external systems.
The external systems can be brought up in difference ways. At this point, I think we might need to support two cases:
1. In-memory systems such as in-memory Kafka brokers. 
2. Setup via TestContainers like `KafkaContainer`.
The current framework seems working fine with both cases, but interface wise we can make it more explicit.

### The environment to run Flink jobs.
1. For IT cases, the environment would likely be a mini-cluster.
2. For E2E tests, the environment would be a Flink TestContainer.

The current testing framework seems missing the IT cases support.

### A testing suite with the testing methods that can be shared across the all the connectors.
For a test case, be it IT case or E2E case, the developers usually changes two things for different tests:
1. The configurations of the job / source / sink.
2. The scenarios, e.g. JM failure / TM Failure / SavePoint + Restore / Change parallelism / Network failure, etc. These scenarios can run against different configurations of the jobs / Source / Sink configurations, thus orthogonal to the (1).

So ideally the testing framework should help developers easily write a test for specific config + scenarios. That indicates something like:

```
public static void testScenario1(TestContext testContext, StreamExecutionEnvironment sEnv)

public static void testScenario2(TestContext testContext, StreamExecutionEnvironment sEnv)

public static void testScenario3(TestContext testContext, StreamExecutionEnvironment sEnv)

...
```

PatrickRen commented:
Hi @becketqin ,

Thank you very much for your review~ I refactored the testing framework based on your comments, and I added a README file for describing the top-level design of the new framework. 

About those comments you left above, I fixed most of them except the hard-coded host and port of the ControllableSource. I think it's necessary to keep the port number a fixed value under some scenarios. For example if I'm using ControllableSource in Testcontainers, I have to know exactly what the port number is so that I can expose this port out of the container. But yes your concern about conflict is indeed reasonable. Do you have any suggestions of solving this? 

Another problem is the incompatibility between different Flink versions (specifically 1.11.2 and current 1.12-SNAPSHOT on master). Docker repository only has our released versions of images like 1.11.2. If I use ```StreamExecutionEnvironment.execute()``` in 1.12-SNAPSHOT to submit the job to 1.11.2 cluster on Docker, the job will be failed because of an NPE from the operator. So currently the ```KafkaE2ETest``` will fail because of this incompatibility. Although I can use a 1.11.2 ```StreamExecutionEnvironment``` to make the case work, it'll be weird always using a version behind the latest one. 

Looking forward to your advices~

PatrickRen commented:
Hi @becketqin , I pushed several commits again for wrapping things up. 

About those two issues I mentioned above, I have solved the Docker image issue by adding a shell script for building a Docker image based on current branch, and this script will be run at the pre-integration-test phase in the Maven lifecycle. 

For the issue about port conflicting, I didn't find any satisfying solution for now, so the `ControllableSource` will only work without any parallelism. I'll try to solve this issue by developing a network proxy in future iterations, and this proxy can also be used for introducing network failures. Currently users can choose other kind of sources for testing parallelism. 

Can you please have a look at my new implementations? Thanks~
